Allow `groupBy` to recreate already closed substreams #24758

markusthoemmes opened the pull request:
`groupBy`'s default behavior is to filter elements, that would go to a substream that is already closed. This allows `groupBy` to recreate those already closed streams, if wanted, and run them as if they would've never run before in the first place.

Closes #24758

ktoso commented:
OK TO TEST

ktoso commented on akka-docs/src/main/paradox/stream/stream-substream.md:
```diff
@@ -26,6 +26,8 @@ This operation splits the incoming stream into separate output
 streams, one for each element key. The key is computed for each element
 using the given function, which is `f` in the above diagram. When a new key is encountered for the first time
 a new substream is opened and subsequently fed with all elements belonging to that key.
+If `allowClosedSubstreamRecreation` is set to `true` a substream belonging to a specific key
```
nice name for it :)

ktoso commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowGroupBySpec.scala:
```diff
@@ -459,6 +459,34 @@ class FlowGroupBySpec extends StreamSpec {
+      // Cleanup, not part of the actual test
+      up.sendComplete()
+      down.expectComplete()
+    }
```
nice clean test

ktoso commented on akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala:
```diff
@@ -214,7 +214,7 @@ import scala.collection.JavaConverters._
  * INTERNAL API
  */
-@InternalApi private[akka] final class GroupBy[T, K](val maxSubstreams: Int, val keyFor: T ⇒ K) extends GraphStage[FlowShape[T, Source[T, NotUsed]]] {
+@InternalApi private[akka] final class GroupBy[T, K](val maxSubstreams: Int, val keyFor: T ⇒ K, val allowClosedSubstreamRecreation: Boolean = false) extends GraphStage[FlowShape[T, Source[T, NotUsed]]] {
```
good 👍

ktoso commented on akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala:
```diff
@@ -334,8 +334,9 @@ import scala.collection.JavaConverters._
     override protected def onTimer(timerKey: Any): Unit = {
```
should we also do a `closedSubstreams = Collections.unmodifiable(Set.empty)` in the `preStart` so no one accidentally uses it?

markusthoemmes replied:
Why not add it in the initialization directly? Any concerns?

ktoso replied:
Yeah thats good too

ktoso commented on akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala:
```diff
@@ -1707,7 +1707,9 @@ final class Flow[In, Out, Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends Gr
    * using the given function. When a new key is encountered for the first time
    * a new substream is opened and subsequently fed with all elements belonging to
-   * that key.
+   * that key. If `allowClosedSubstreamRecreation` is set to `true` a substream
```
Good text, though make it an `@param allowClosedSubstreamRecreation` in a new line

ktoso commented on akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala:
```diff
@@ -1743,8 +1745,20 @@ final class Flow[In, Out, Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends Gr
+   * using the given function. When a new key is encountered for the first time
+   * a new substream is opened and subsequently fed with all elements belonging to
+   * that key.
+   *
```
let's add docs here that the keys "closed" cause those values to be later on drained directly, without sending into the subflow, and that that set is never cleared -- so if infinite keys are expected it could be unsafe memory wise (put a WARNING: ....)

ktoso commented on akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:
```diff
@@ -1802,15 +1804,15 @@ trait FlowOps[+Out, +Mat] {
    *        that are supported; if more distinct keys are encountered then the stream fails
    */
-  def groupBy[K](maxSubstreams: Int, f: Out ⇒ K): SubFlow[Out, Mat, Repr, Closed] = {
+  def groupBy[K](maxSubstreams: Int, f: Out ⇒ K, allowClosedSubstreamRecreation: Boolean = false): SubFlow[Out, Mat, Repr, Closed] = {
```
also as the override please; default methods are hard to evolve in a bin compat way if we ever have to add another overload

markusthoemmes replied:
You mean: as the overload? As in: don't use default params but provide an overloaded method?

markusthoemmes replied:
Makes sense, I've been bitten by those often enough. I agree it can become hard to evolve, especially if you start to need overloads with a differing parameter set. Will change!

ktoso reviewed (approved):
A few minor improvements; exactly the solution to the ticket I was hoping for, thanks a lot!

markusthoemmes committed:
Allow `groupBy` to recreate already closed substreams #24758

`groupBy`'s default behavior is to filter elements, that would go to a substream
that is already closed. This allows `groupBy` to recreate those already closed streams,
if wanted, and run them as if they would've never run before in the first place.

Overload instead of default parameter

Make set of closed keys unmodifiable if not needed

Adjust documentation, adding a warning on memory consumption.

Add MiMa exclude.

markusthoemmes commented:
@ktoso Adjusted for all comments. I did not put a warning into the substream docs, it seems like that docs are more geared towards actual behavior vs. edge-cases. WDYT?

ktoso commented on akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala:
```diff
@@ -1707,9 +1707,12 @@ final class Flow[In, Out, Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends Gr
+   * keeps track of all keys of streams that have already been closed. If you expect an infinite
+   * number of keys this can cause memory issues. Elements belonging to those keys are drained
+   * directly and not send to the substream.
    *
```
please also add the warning to the normal, without the boolean param, method; since there this WARNING behavior will always be the one

markusthoemmes replied:
Ah sure, that makes lots of sense! Changed and pushed!

ktoso commented on akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala:
```diff
@@ -224,7 +225,7 @@ import scala.collection.JavaConverters._
     lazy val decider = inheritedAttributes.mandatoryAttribute[SupervisionStrategy].decider
     private val activeSubstreamsMap = new java.util.HashMap[Any, SubstreamSource]()
-    private val closedSubstreams = new java.util.HashSet[Any]()
+    private val closedSubstreams = if (allowClosedSubstreamRecreation) Collections.unmodifiableSet(new java.util.HashSet[Any]) else new java.util.HashSet[Any]()
```
👍

ktoso commented:
One additional docs change and we're lookin good here

markusthoemmes commented:
Seems like the additional issue is with regards to binary compatibility, but on an `@InternalApi`, so I guess I'm good from that PoV?

ktoso commented on akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:
```diff
@@ -1801,20 +1806,38 @@ trait FlowOps[+Out, +Mat] {
+   * a new substream is opened and subsequently fed with all elements belonging to
+   * that key.
+   *
+   * WARNING: The stage keeps track of all keys of streams that have already been closed.
```
The big warning we should also add to the operators/groupBy.md page.

markusthoemmes replied:
Should it be one of the big red callout blocks or is a "Note" sufficient for the docs?

ktoso commented on akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala:
```diff
@@ -224,7 +225,7 @@ import scala.collection.JavaConverters._
     lazy val decider = inheritedAttributes.mandatoryAttribute[SupervisionStrategy].decider
     private val activeSubstreamsMap = new java.util.HashMap[Any, SubstreamSource]()
-    private val closedSubstreams = new java.util.HashSet[Any]()
+    private val closedSubstreams = if (allowClosedSubstreamRecreation) Collections.unmodifiableSet(new java.util.HashSet[Any]) else new java.util.HashSet[Any]()
```
`Collections.unmodifiableSet(Collections.emptySet) ` this is the case where we never want to touch this collection -- this also means we'll never "infinitely accumulate keys"

markusthoemmes replied:
Is this about changing documentation or "just" change the code to your suggestion?

ktoso replied:
Actually just to `Collections.emptySet` since it is immutable already :) yes just code

ktoso commented:
> Seems like the remaining issue is with regards to binary compatibility, but on an @InternalApi, so I guess I'm good from that PoV?

Yeah but you will have to add it to excludes, please see CONTRIBUTING.md about binary compatibility checks and MiMa :)

ktoso commented:
had a minor comment i just noticed

patriknw commented on akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala:
```diff
@@ -1709,6 +1709,11 @@ final class Flow[In, Out, Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends Gr
+   * WARNING: If `allowClosedSubstreamRecreation` is set to `false` (default behavior) the stage
+   * keeps track of all keys of streams that have already been closed. If you expect an infinite
+   * number of keys this can cause memory issues. Elements belonging to those keys are drained
+   * directly and not send to the substream.
```
Shouldn't we also warn about the semantics of `allowClosedSubstreamRecreation=true` as discussed in the other PR:

https://github.com/akka/akka/pull/24783#issuecomment-376769171

https://github.com/akka/akka/pull/24783#issuecomment-376794466

markusthoemmes replied:
Hmm, at least a comment on the racyness would be appropriate I think. I had to circumvent that problem in the tests as well.

```scala
Source(1, 2, 1, 1).groupBy(Int.MaxValue, identity, true).take(1).mergeSubstreams.runForeach(println)
```

Isn't guaranteed to print all the elements for instance.

huntc commented on akka-docs/src/main/paradox/stream/operators/Source-or-Flow/groupBy.md:
```diff
@@ -14,7 +14,21 @@ Demultiplex the incoming stream into separate output streams.
+
+If `allowClosedSubstreamRecreation` is set to `false` (default behavior) the stage keeps track of all
+keys of streams that have already been closed. If you expect an infinite number of keys this can cause
+memory issues. Elements belonging to those keys are drained directly and not send to the substream.
```
PR https://github.com/akka/akka/pull/24831 prevents an infinite amount of keys and could be applied to `allowClosedSubstreamRecreation` being set to `false`.

huntc commented on akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala:
```diff
@@ -1742,9 +1751,28 @@ final class Flow[In, Out, Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends Gr
+   *
+   * @see [[#groupBy]]
    */
   def groupBy[K](maxSubstreams: Int, f: function.Function[Out, K]): SubFlow[In, Out, Mat] =
```
Now we have two `groupBy` operations that are semantically the same when `allowClosedSubstreamRecreation` is set to `false`. Understandably, we do this to retain binary compatibility. However, should we deprecate this method in favour of the other so that it can be removed at a later date?

patriknw replied:
depends on if we want to bias towards one as the default, and I assume that the `allowClosedSubstreamRecreation=false` has the least surprising semantics given what Roland explained so it's still good to have as the default

patriknw commented:
I rebased and moved the mima filters

patriknw reviewed (approved):
LGTM

huntc commented:
I'm not entirely familiar with the Java DSL structure, but it seems as though there's a `Flow` and a `Source`, and that this fix didn't make it to the `Source`. Have I read this correctly? I can raise an issue if so.

ktoso commented:
I think you're right, it's missing; weird that the DlsConsistencySpec did not catch this :(
Please PR, thanks in advance!
